src/widget_chooser.c

```diff
@@ -0,0 +1,502 @@
+
+void widget_chooser_save(variable *var)
+{
+	FILE             *outfile;
+	GList            *element;
+	GtkTextBuffer    *buffer;
+	GtkTextIter       start, end;
+	gchar            *act;
+	gchar            *filename = NULL;
+	gchar            *text;
+	gchar            *string;
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Entering.\n", __func__);
+#endif
+
+	/* We'll use the output file filename if available */
+	act = attributeset_get_first(&element, var->Attributes, ATTR_OUTPUT);
+	while (act) {
+		if (strncasecmp(act, "file:", 5) == 0 && strlen(act) > 5) {
+			filename = act + 5;
+			break;
+		}
+		act = attributeset_get_next(&element, var->Attributes, ATTR_OUTPUT);
+	}
+
+	/* If we have a valid filename then open it and dump the widget's data to it */
+	if (filename) {
+		text = gtk_file_chooser_get_filename(GTK_FILE_CHOOSER(var->Widget));
+			if (text) {
+				string = g_strdup(text);
+				if ((outfile = fopen(filename, "w"))) {
+					fprintf(outfile, "%s", string);
+					fclose(outfile);
+				} else {
+					fprintf(stderr, "%s(): Couldn't open '%s' for writing.\n",
+						__func__, filename);
+				}
+				g_free(text);
+			}
+	} else {
+		fprintf(stderr, "%s(): No <output file> directive found.\n", __func__);
+	}
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Exiting.\n", __func__);
+#endif
+}
+
+/***********************************************************************
+ * Input by Command                                                    *
+ ***********************************************************************/
+
+static void widget_chooser_input_by_command(variable *var, char *command)
+{
+	FILE             *infile;
+	GString          *text = g_string_sized_new(512);
+	gchar             line[512];
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Entering.\n", __func__);
+#endif
+
+#ifdef DEBUG_CONTENT
+	fprintf(stderr, "%s(): command: '%s'\n", __func__, command);
+#endif
+
+	/* Opening pipe for reading... */
+	if (infile = widget_opencommand(command)) {
+		/* Read the file one line at a time */
+		while (fgets(line, 512, infile)) {
+			g_string_append(text, line);
+		}
+
+		gtk_file_chooser_set_filename(GTK_FILE_CHOOSER(var->Widget), text->str);
+		g_string_free(text, TRUE);
+
+		/* Close the file */
+		pclose(infile);
+	} else {
+		fprintf(stderr, "%s(): Couldn't open '%s' for reading.\n", __func__,
+			command);
+	}
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Exiting.\n", __func__);
+#endif
+}
+
+/***********************************************************************
+ * Input by File                                                       *
+ ***********************************************************************/
+
+static void widget_chooser_input_by_file(variable *var, char *filename)
+{
+	FILE             *infile;
+	GString          *text = g_string_sized_new(512);
+	gchar             line[512];
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Entering.\n", __func__);
+#endif
+
+	if (infile = fopen(filename, "r")) {
```

Comment:
What if the input file is not a regular file? For example, a FIFO, or a special file in /proc. fgets() can get stuck forever.

Reply:
It would get stuck, yes, but this is throughout gtkdialog's code - it was designed to only use regular files as signal inputs. To change this we'd have to rewrite all widgets.